src/archivematicaCommon/lib/archivesspace/client.py

```diff
@@ -0,0 +1,424 @@
+    pass
+
+
+class AuthenticationError(Exception):
+    pass
+
+
+class CommunicationError(Exception):
+    def __init__(self, status_code, response):
+        message = "ArchivesSpace server responded {}".format(status_code)
+        self.response = response
+        super(CommunicationError, self).__init__(message)
+
+
+class ArchivesSpaceError(Exception):
+    pass
+
+
+class ArchivesSpaceClient(object):
+    """
+    Client to communicate with a remote ArchivesSpace installation using its backend API.
+
+    Note that, while functions follow the same API as the ArchivistsToolkitClient, one major difference is the handling of resource and component IDs.
+    In ArchivistsToolkitClient, resource IDs are longs representing the database row ID.
+    In this client, resource IDs are instead URI fragments representing the location of the record, for instance:
+        /repositories/2/resource/1
+    This change is due to the fact that the integer IDs are not unique across the collection in ArchivesSpace, as they were in Archivist's Toolkit.
+    """
+
+    RESOURCE = 'resource'
+    RESOURCE_COMPONENT = 'resource_component'
+
+    def __init__(self, host, user, passwd, port=8089, repository=2):
+        parsed = urlparse(host)
+        if not parsed.scheme:
+            host = 'http://' + host
+
+        self.host = host + ':' + str(port)
+        self.user = user
+        self.passwd = passwd
+        self.repository = '/repositories/{}'.format(repository)
+        self._login()
+
+    def _login(self):
+        try:
+            response = requests.post(self.host + '/users/' + self.user + '/login',
+                                     data={'password': self.passwd})
+        except requests.ConnectionError as e:
+            raise ConnectionError("Unable to connect to ArchivesSpace server: " + str(e))
+
+        try:
+            output = response.json()
+        except:
+            raise ArchivesSpaceError("ArchivesSpace server responded with status {}, but returned a non-JSON document".format(response.status_code))
+
+        if 'error' in output:
+            raise AuthenticationError("Unable to log into ArchivesSpace installation; message from server: {}".format(output['error']))
+        else:
+            token = output['session']
+
+        self.session = requests.Session()
+        self.session.headers.update({'X-ArchivesSpace-Session': token})
+
+    def _request(self, method, url, params, expected_response, data=None, retry=True):
+        if not url.startswith('/'):
+            url = '/' + url
+
+        response = method(self.host + url, params=params, data=data)
+        if response.status_code != expected_response:
+            # session has expired; acquire a new session, then retry once
+            if retry and response.status_code == 412 and "SESSION_GONE" in response.text:
+                self._login()
+                return self._request(method, url, params, expected_response, data=data, retry=False)
```

Comment:
It recurses with the `retry` kwarg set to `False`, and L78 checks that first - it will never retry more than once.